File: sorts/sorts/merge_sort/merge_sort.hs
-- This Haskell program contains the implementation of a merge sort algorithm.

-- This function sorts an array from least to greatest using the
-- merge sort algorithm.
merge_sort :: (Ord a) => [a] -> [a]
merge_sort [] = []
merge_sort [x] = [x] 
merge_sort list = merge (merge_sort $ fst $ halve list)
                        (merge_sort $ snd $ halve list)
  where
    -- This function splits a list into two halves.
    halve :: [a] -> ([a], [a])
    halve l = splitAt (div (length list) 2) l
    -- This function merges two lists into one list in ascending order,
    -- assuming the original lists are in increasing order.
    merge :: (Ord a) => [a] -> [a] -> [a]
    merge l1 l2
      | l1 == []          = l2
      | l2 == []          = l1
      | head l1 < head l2 = [head l1] ++ merge (tail l1) l2
      | otherwise         = [head l2] ++ merge l1 (tail l2)

File: sorts/sorts/merge_sort/merge_sort.erl
% This Erlang module sorts a list of values using the merge sort algorithm.

-module(merge_sort).
-export([
	clear/0,
	merge_sort/1
]).

% This function clears the interpreter screen.
clear() -> io:format(os:cmd(clear)).

% This function merges two lists which are already sorted into a single
% resulting sorted list.
merge([], L) -> L;
merge(L, []) -> L;
merge([X|Xs], [Y|Ys]) ->
	case X < Y of
		true  -> [X] ++ merge(Xs, [Y]++Ys);
		false -> [Y] ++ merge([X]++Xs, Ys)
	end.

% This function returns the middle index of a given list.
list_half_index(L) -> trunc(length(L)/2).

% This function uses the merge sort algorithm to sort a list.
merge_sort([]) -> [];
merge_sort([X]) -> [X];
merge_sort(L) ->
	{L1, L2} = lists:split(list_half_index(L), L),
	merge(merge_sort(L1), merge_sort(L2)).